File: src/com/company/simple/datastruct/Trie.java
package com.company.simple.datastruct;

/**
 * @description: notcallme
 * @author: 15262
 * @date: 2020/3/28
 */

public class Trie {
    TrieNode root;

    /**
     * Initialize your data structure here.
     */
    public Trie() {
        this.root = new TrieNode();
    }

    /**
     * Inserts a word into the trie.
     */
    public void insert(String word) {
        TrieNode cur = root;
        for (int i = 0; i < word.length(); i++) {
            int c = word.charAt(i) - 'a';
            if (cur.children[c] == null) {
                cur.children[c] = new TrieNode(c);
            }
            cur = cur.children[c];
        }
        cur.isEnd = true;
    }

    /**
     * Returns if the word is in the trie.
     */
    public boolean search(String word) {
        TrieNode cur = root;
        for (int i = 0; i < word.length(); i++) {
            int val = word.charAt(i) - 'a';
            if (cur.children[val] == null) {
                return false;
            }
            cur = cur.children[val];
        }
        return cur.isEnd;
    }

    /**
     * Returns if there is any word in the trie that starts with the given prefix.
     */
    public boolean startsWith(String prefix) {
        TrieNode cur = root;
        for (int i = 0; i < prefix.length(); i++) {
            int val = prefix.charAt(i) - 'a';
            if (cur.children[val] == null) {
                return false;
            }
            cur = cur.children[val];
        }
        return true;
    }
}

File: 2021/04/Trie.go
package main

import "fmt"

type Trie struct {
	val  rune
	end  bool
	next []*Trie
}

/** Initialize your data structure here. */
func Constructor() Trie {
	return Trie{
		next: make([]*Trie, 26),
	}
}

/** Inserts a word into the trie. */
func (this *Trie) Insert(word string) {
	node := this
	for _, c := range word {
		v := c - 'a'
		if node.next[v] == nil {
			trie := Constructor()
			node.next[v] = &trie
			node.next[v].val = c
		}
		node = node.next[v]
	}

	node.end = true
}

/** Returns if the word is in the trie. */
func (this *Trie) Search(word string) bool {
	node := this
	for _, c := range word {
		v := c - 'a'
		if node.next[v] == nil {
			return false
		}
		node = node.next[v]
	}
	return node.end
}

/** Returns if there is any word in the trie that starts with the given prefix. */
func (this *Trie) StartsWith(prefix string) bool {
	node := this
	for _, c := range prefix {
		v := c - 'a'
		if node.next[v] == nil {
			return false
		}
		node = node.next[v]
	}
	return true
}

/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor()
 * obj.Insert(word)
 * param_2 := obj.Search(word)
 * param_3 := obj.StartsWith(prefix)
 */

func main() {
	trie := Constructor()
	trie.Insert("apple")
	fmt.Println(trie.Search("apple"))   // 返回 True
	fmt.Println(trie.Search("app"))     // 返回 False
	fmt.Println(trie.StartsWith("app")) // 返回 True
	trie.Insert("app")
	fmt.Println(trie.Search("app")) // 返回 True
}
